common/utils/src/main/scala/org/apache/spark/util/ClosureCleaner.scala

```diff
@@ -499,72 +503,45 @@ private[spark] object ClosureCleaner extends Logging {
-        }
-      val outerClone = cmdHelperOuter.flatMap(o => cmdClones.get(o.getClass)).orNull
-      logDebug(s" + Cloning instance of Ammonite command helper class ${cmdHelperClass.getName}")
-      cmdClones(cmdHelperClass) =
-        instantiateClass(cmdHelperClass, enclosingObject = outerClone)
-    }
+    // Instead of cloning and trying to update lambda reference, directly modify the original
+    // REPL object to null out unused fields. This works because REPL objects are not hidden
+    // classes and their fields can be modified.
+    val capturingClass = outerThis.getClass
+    if (accessedFields(capturingClass).size < capturingClass.getDeclaredFields.length) {
+      val fieldsToNull = capturingClass.getDeclaredFields.filterNot { field =>
+        accessedFields(capturingClass).contains(field.getName)
+      }
 
-    // set accessed fields
-    for ((_, cmdClone) <- cmdClones) {
-      val cmdClass = cmdClone.getClass
-      val accessedFields = accessedAmmCmdFields(cmdClass)
-      for (field <- cmdClone.getClass.getDeclaredFields
-           // outer fields were initialized during clone construction
-           if accessedFields.contains(field.getName) && field.getName != "$outer") {
-        // get command clone if exists, otherwise use an original field value
-        val value = cmdClones.getOrElse(field.getType, {
+      for (field <- fieldsToNull) {
+        try {
           field.setAccessible(true)
-          field.get(ammCmdInstances(cmdClass))
-        })
-        setFieldAndIgnoreModifiers(cmdClone, field, value)
+          field.set(outerThis, null)
+        } catch {
+          case _: Exception =>
+            // Ignore failures to set fields - this is a best-effort cleanup
+        }
       }
     }
 
-    val outerThisClone = if (!isAmmoniteCommandOrHelper(outerThis.getClass)) {
-      // if outer class is not Ammonite helper / command object then is was not cloned
-      // in the code above. We still need to clone it and update accessed fields
-      logDebug(s" + Cloning instance of lambda capturing class ${outerThis.getClass.getName}")
-      val clone = cloneAndSetFields(parent = null, outerThis, outerThis.getClass, accessedFields)
-      // making sure that the code below will update references to Ammonite objects if they exist
-      for (field <- outerThis.getClass.getDeclaredFields) {
-        field.setAccessible(true)
-        cmdClones.get(field.getType).foreach { value =>
-          setFieldAndIgnoreModifiers(clone, field, value)
+    // Also clean up Ammonite command fields if any
+    for ((cmdClass, cmdInstance) <- ammCmdInstances) {
+      val cmdAccessedFields = accessedAmmCmdFields.getOrElse(cmdClass, Set.empty)
+      if (cmdAccessedFields.size < cmdClass.getDeclaredFields.length) {
+        val fieldsToNull = cmdClass.getDeclaredFields.filterNot { field =>
+          cmdAccessedFields.contains(field.getName) || field.getName == "$outer"
+        }
+
+        for (field <- fieldsToNull) {
```

Comment:
It seems that each modification has similar code. Can we extract a common function?